### ISSUE TITLE
Translatable and inheritance mapping

### ISSUE THREAD

**Issue by jacekpl** [KnpLabs/DoctrineBehaviors#307]
Hi,

I've got two entities: TableOfContents and ContentType. Both of them have also Translatable entities.

```
<?php
namespace AppBundle\Entity;

use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\ORM\Mapping as ORM;
use Knp\DoctrineBehaviors\Model as ORMBehaviors;

/**
 * Class TableOfContents
 * @package AppBundle\Entity
 * @ORM\Entity()
 * @ORM\Table()
 * @ORM\InheritanceType("SINGLE_TABLE")
 * @ORM\DiscriminatorColumn(name="type", type="string")
 * @ORM\DiscriminatorMap({"tableOfContents" = "TableOfContents", "pdf" = "PdfType", "content" = "ContentType"})
 */
class TableOfContents
{
    use ORMBehaviors\Translatable\Translatable;

    /**
     * @var int
     * @ORM\Id()
     * @ORM\GeneratedValue(strategy="AUTO")
     * @ORM\Column(type="integer")
     */
    protected $id;

    public function getId()
    {
        return $this->id;
    }
}
```

```
<?php
namespace AppBundle\Entity;

use Doctrine\ORM\Mapping as ORM;
use Knp\DoctrineBehaviors\Model as ORMBehaviors;

/**
 * Class TableOfContentsTranslation
 * @package AppBundle\Entity
 * @ORM\Entity()
 */
class TableOfContentsTranslation
{
    use ORMBehaviors\Translatable\Translation;

    /**
     * @var string
     * @ORM\Column(type="string", length=255)
     */
    protected $title;

    public function getTitle()
    {
        return $this->title;
    }

    public function setTitle($title)
    {
        $this->title = $title;
    }
}
```

```
<?php
namespace AppBundle\Entity;

use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\ORM\Mapping as ORM;
use Knp\DoctrineBehaviors\Model as ORMBehaviors;

/**
 * Class ContentType
 * @package AppBundle\Entity
 * @ORM\Entity()
 * @ORM\HasLifecycleCallbacks()
 */
class ContentType extends TableOfContents
{
    use ORMBehaviors\Translatable\Translatable;

    const TYPE = "content";
    //(...)
}
```

```
<?php
namespace AppBundle\Entity;

use Doctrine\ORM\Mapping as ORM;
use Knp\DoctrineBehaviors\Model as ORMBehaviors;

/**
 * Class ContentTypeTranslation
 * @package AppBundle\Entity
 * @ORM\Entity()
 */
class ContentTypeTranslation
{
    use ORMBehaviors\Translatable\Translation;

    /**
     * @var string
     * @ORM\Column(type="string", length=10000)
     */
    private $content;

    public function getContent()
    {
        return $this->content;
    }

    public function setContent($content)
    {
        $this->content = $content;
    }
}
```

I would like to translate title field for ContentType object.

When I try to do:
```
$content = new ContentType();
$content->translate('en')->setTitle('test');
```

I get `Fatal error: Call to undefined method AppBundle\Entity\ContentTypeTranslation::setTitle() `

How to make it work? I tried to extend TableOfContentsTranslation by ContentTypeTranslation, but it doesn't work either. I haven't found any example or description on how to use translatable with entities inheritance. 

I would appreciate any help:)

**Comment by taylankasap** [comment]
What was the problem when you

> tried to extend TableOfContentsTranslation by ContentTypeTranslation

**Comment by jbenezech** [comment]
Extending the translation class seems to work alright but requires a bit of fiddling with doctrine migrations.
doctrine:migration:diff produces an error:
```There is no column with name 'translatable_id' on table 'member_translation'.```

However, after migrating manually, the app seems to work fine.

**Comment by jbenezech** [comment]
Digging a bit further, the problem comes from the TranslatableSubscriber adding a unique constraint on the translation subclass :

```
private function mapTranslation(ClassMetadata $classMetadata)
    {
...
    $name = $classMetadata->getTableName().'_unique_translation';
    if (!$this->hasUniqueTranslationConstraint($classMetadata, $name)) {
        $classMetadata->table['uniqueConstraints'][$name] = [
            'columns' => ['translatable_id', 'locale' ]
        ];
    }
```
In my case, using JOINED inheritance, the table of the subclass doesn't have a translatable_id or locale column. Therefore the createIndex fails when generating the migration.

I'm not sure this unique index should be added at all even for other types of inheritance since it would be created by generating the migration of the super class anyways.

Couldn't figure out a clever workaround so for now resorted to hack the subscriber this way

```
if (!$this->hasUniqueTranslationConstraint($classMetadata, $name) &&
    $classMetadata->getName() == $classMetadata->rootEntityName) {
    $classMetadata->table['uniqueConstraints'][$name] = [
        'columns' => ['translatable_id', 'locale' ]
    ];
}
```

**Comment by jbenezech** [comment]
Duplicated: https://github.com/KnpLabs/DoctrineBehaviors/pull/317

**Comment by docteurklein** [comment]
closed by #317
